### ISSUE TITLE
Add a unprovision job in order to clean configuration previously loaded by the import job

### ISSUE THREAD

**Issue by JCL38-ORANGE** [starkandwayne/shield-boshrelease#156]
Hello,

My BOSH deployment is using shield v8 with the mongodb bosh release => https://github.com/orange-cloudfoundry/mongodb-boshrelease.
Each mongodb node (instance_group) has a shield agent job with a proper configuration loaded with the errand import job.
If an operator requires a scale down (node removal), there is actually no way to clean configuration loaded previously with the import job.

Thanks and regards,
Jean-Christophe.

**Comment by patrick1378** [comment]
Hey Jean-Christophe, 
Could I possibly see your import just for some context on your agent? Thanks.
Pat

**Comment by JCL38-ORANGE** [comment]
Hello Pat,

You can find below the configuration for the import job : 

- type: replace
  path: /instance_groups/name=mongod/jobs/-
  value:
    name: import
    release: shield
    properties:
      generate_token: true
      core: sandbox
      domain: shield-webui-((deployment_name)).((/secrets/cloudfoundry_ops_domain))
      import:
        core: https://shield-webui-((deployment_name)).((/secrets/cloudfoundry_ops_domain))
        insecure_skip_verify: true
        tenants:
        - name: ((deployment_name))

          members:
          - user: admin@local
            role: admin

          storage:
          - name: minio-mongodb
            summary: |
              Internal S3 cloud storage for all SHIELD tenants to use
            agent:  (ip):5444
            plugin: s3
            config:
              s3_host: 192.168.XX.YY
              s3_port: "443"
              access_key_id: ((minio_access_key_id))
              secret_access_key: ((minio_secret_access_key))
              bucket: shieldv8-mongodb #pre-requisite
              skip_ssl_validation: true
          - name: obosv4-mongodb
            summary: |
              Obos v4 cloud storage for all SHIELD tenants to use
            agent:  (ip):5444
            plugin: s3
            config:
              s3_host: ((s3_host))
              s3_port: "443"
              access_key_id: ((access_key_id))
              secret_access_key: ((secret_access_key))
              bucket: shieldv8-mongodb #pre-requisite
              skip_ssl_validation: true

          policies:
          - name: short-term
            days: 7

          systems:
          - name: mongo-fpc-(ip)
            agent: (ip):5444
            config:
              mongo_host: (ip)
              mongo_user: MongoRoot
              mongo_password: ((mongo_root_password))
              mongo_bindir: "/var/vcap/packages/mongodb/bin"
            jobs:
            - name: (ip)-backup-test
              paused: true
              policy: short-term
              storage: minio-mongodb
              when: every 8 hours from 0:15
            plugin: mongo
            summary: shield v8

Thanks and regards,
Jean-Christophe.

**Comment by JCL38-ORANGE** [comment]
Hello Patrick,

In fact we want to clean all tenant configuration objects.
In our case, the (deployment_name) tenant configuration : members, storage, policies and systems must be removed from the system.
Thanks and regards,
Jean-Christophe.

**Comment by patrick1378** [comment]
Hi Jean-Christophe, 
In your environment where are your tenants in relation to shield core? For example, do you have a shield per service broker and a tenant per service instance?

**Comment by JCL38-ORANGE** [comment]
Hi Patrick,

The bosh deployment is the following : 
Instance                                             Process State  AZ  IPs  
mongod/4717c515-ae44-4922-8f3e-0b259f3d5853          running        z1  192.168.211.21  
mongod/5540930f-3dd2-49c5-b41b-27be638764da          running        z2  192.168.211.22  
mongod/c4592c15-3ffe-46e0-a76a-b596c8d9a0f0          running        z3  192.168.211.23  
mongodb-broker/642972f9-e7e2-4953-b69f-1b8311c60b50  running        z1  192.168.211.24  
shield/949a37a6-b7a4-4fd6-9206-eb69439c4a9d          running        z1  192.168.211.25  

shield is the shield core.
shield agents are embedded in mongod.
On this shield core, I have only one tenant named with the deployment_name and three systems (one by node).
Regards,
Jean-Christophe.

**Comment by patrick1378** [comment]
How exactly does your mongo broker work? Thanks again for the help!

**Comment by JCL38-ORANGE** [comment]
Hi Patrick,

The mongodb broker doesn't interact at all with shield.
It implements OSB API lifecycle i..e provisions database and users in mongodb backend and returns credentials once provisioning is done.
Regards,
Jean-Christophe.

**Comment by jhunt** [comment]
We have two options for this feature:

1. Explicitly require a list of all of the things to delete.  This is the safer, more general approach.
2. Re-use the import configuration yaml, and essentially implement `shield unimport`; deleting everything that would have been imported by a `shield import`; this is easier for ops, but more dangerous.

The first option would look something like this:

```
instance_groups:
  - name: decomission
    jobs:
      - name: deprovision
        release: shield
        deprovision:
          SYSTEM:
            - store: Global Storage System
            - tenant: temp-tenant
          tenant1:
            - target: some-database
            - target: files
            - store: Local Storage
            - policy: Short-Term
```

The semantics of this would need to be hashed out; for example, does listing a `tenant` for deprovision under `SYSTEM` cause all things owned by that tenant to be deleted?  My initial answer is "yes", but we need to talk it out.

**Comment by JCL38-ORANGE** [comment]
Hi Patrick and James,

I agree with you. My preference is for the first option because it is a fine grain approach.
The semantic for things under special tenant SYSTEM looks good to me (listing a tenant for deprovision under SYSTEM cause all things owned by that tenant to be deleted).

Regards,
Jean-Christophe.

**Comment by JCL38-ORANGE** [comment]
Hi Patrick and James,

Just one point about the retention policy.
If I define : 
...
policies:
          - name: short-term
            days: 7
...
I expect that the configuration and the backups in S3 will be cleanup (accordingly to the policy defined) in order to avoid orphan/lacks backups in the S3.
Could you confirm that you are going to implement this point also ?
It is a kind of lazy/deferred cleaning.

Thanks and regards,
Jean-Christophe.

**Comment by patrick1378** [comment]
Hi Jean-Christophe, the current plan is to mark the archives as needing to be purged when we are cleaning up the configs from a deleted tenant. This will allow shield to dispose of the archives the similarly to the way it handles expired archives.

**Comment by patrick1378** [comment]
With 238425e8a73dc33a1ca4f4ab3671e58f584f0731, A new errand has been created to run the new delete-tenant shield cli command.
